### ISSUE TITLE
Create a unified networking API.

### ISSUE THREAD

**Issue by kozyraki** [mesosphere/marathon#1112]
The Mesos protobufs have been extended with basic service discovery info for tasks (see https://reviews.apache.org/r/29473/ for details). This information can be used by systems Mesos-DNS to generate DNS records for service discovery. 

To facilitate apps that use multiple ports, it would be great if Marathon allowed the user to specify a name and a protocol when requesting ports. For example, when specifying task X, instead of typing:
"ports": [
    80,
    443]
we could have:
"ports": [
    80:http:tcp,
    443:admin:*]
with the two fields after the port number being optional. This information will be translated by Marathon the ports  task info in https://reviews.apache.org/r/29473/  and will lead to the generation of SRV records for names like:
   _http.tcp.x.domain
  _admin.tcp.x.domain
  _admin.udp.x.domain

There will also be cases where the users does not care about the specific ports but will want to name the ports assigned by Marathon. For instance, if the task is launched as "/usr/local/bin/X -p1 $PORT1 -p2 $PORT2", we can allow a syntax like:
"ports": [
    port1:http:tcp,
    port2:admin:*]
### Sub-Tasks
- #670 - Rethink ports API 
- #1866 - Set DiscoveryInfo for created TaskInfo messages

**Comment by mindscratch** [comment]
Having named ports would be helpful, in particular for supporting DNS tools such as mesos-dns. Kubernetes provides a [name](https://godoc.org/github.com/GoogleCloudPlatform/kubernetes/pkg/api#Port) for ports which has also been helpful for a similar use case.

**Comment by rasputnik** [comment]
A general way to add app metadata would be invaluable; I was planning to take advantage of the new app labels (key/value fields) in 0.8.0 to store this - see https://github.com/mesosphere/marathon/pull/1051 - could meson-dns etc. reuse that feature to store the information it needs?

I know you Scala folks love your strong typing :) , but this seems the most flexible option.

**Comment by tnachen** [comment]
I also think the need for a more generic port structure is desired too, as there was
another PR that was needing a way to reference Marathon allocated port.

I wonder if it's possible to even generalize this and introduce a port struct?

Ports:[
{ "port"; 80,
   "name": "http"
   "protocol": "tcp" },
{ "port": 0,
  "name": "internal",
  "protocol": "tcp" }]

Note that port 0 will be random assigned marathon based on offer.

IMO this is a lot more flexible and more
clear of the intention.

**Comment by kozyraki** [comment]
@air Here is the issue about naming ports.

**Comment by air** [comment]
Looks like a solid addition with precedent in k8s. I'd like to see discussion on the tradeoffs between [using labels](https://github.com/mesosphere/marathon/issues/1112#issuecomment-71793490) and using a [new struct](https://github.com/mesosphere/marathon/issues/1112#issuecomment-71794443).

**Comment by kozyraki** [comment]
@air from my point of view either approach is ok as long as Marathon translates properly to the ServiceDiscovery Info in the task info!

**Comment by kolloch** [comment]
@drexin and @ConnorDoyle should definitely voice their opinion on this :)

Actually, there is some confusion around the current role of the ports field in Marathon in general. @tnachen hinted at that already.

In the case of docker containers, it is possible to differentiate between 
* service port
* container port
* host port

For the mesos executor, there is a service port and a host port but you cannot specify them seperately. The service port is only needed with port based service discovery. I guess that mesos-dns should be preferred to that anyhow.

So I am in favor of the struct-based approach to model this issue into one API.

It might technically be difficult to use the same field name with different types, so we might have to change the field name (favorites?).

# Examples

## One dynamic port without service discovery

An application with an automatically allocated port without any service discovery. No globally unique service port will be allocated. The started task gets the dynamically allocated port as `$PORT` and `$PORT0` variable.
```json
"portDefs": [
  {
    "hostPort": 0
  }
]
```

## One dynamic port with DNS SRV service discovery

An application with an automatically allocated port. No globally unique service port will be allocated. The started task gets the dynamically allocated port as `$PORT` and `$PORT0` variable. A SRV record for `_http.tcp.x.domain` will be created by `mesos-dns`.
```json
"portDefs": [
  {
    "hostPort": 0,
    "name": "http",
    "protocol": "tcp"
  }
]
```

## Multiple dynamic ports with DNS SRV service discovery

An application with an automatically allocated port. No globally unique service port will be allocated. The started task gets the dynamically allocated ports as `$PORT0` and `$PORT1` variables. SRV records for `_http.tcp.x.domain` and `_metrics.tcp.x.domain` will be created by `mesos-dns`.
```json
"portDefs": [
  {
    "hostPort": 0,
    "name": "http",
    "protocol": "tcp"
  },
  {
    "hostPort": 0,
    "name": "metrics",
    "protocol": "tcp"
  }
]
```

## One dynamic port with port-based service discovery

An application with an automatically allocated port. A globally unique service port will be allocated. The started task gets the dynamically allocated port as `$PORT` and `$PORT0` variable. 
```json
"portDefs": [
  {
    "hostPort": 0,
    "servicePort": 0
  }
]
```

## One dynamic task port with port-based service discovery on a well-known port

An application with an automatically allocated port. A tool can use the supplied service port. The started task gets the dynamically allocated port as `$PORT` and `$PORT0` variable. 
```json
"portDefs": [
  {
    "hostPort": 0,
    "servicePort": 8080
  }
]
```

**Comment by ConnorDoyle** [comment]
Picking this back up after a long hiatus!  A goal associated with service discovery is isolation via external overlay networks.  Just had a chat with @kozyraki offline and we came up with an initial proposal.  The ports part is similar to @kolloch's approach above, but adds some additional things.

_Goals:_
- To allow users to set Mesos service discovery metadata.
- To allow coarse whitelisting for inbound connections to application endpoints.
- To provide a uniform interface to current Marathon networking options.
  - To deprecate and eventually remove `app.ports`.
  - To deprecate and eventually remove `app.container.docker.network.portMappings`.

```javascript
{

  ...                                // The other app definition fields are elided here for brevity.

  "network": {
    "addresses": {
      "mode": "CONTAINER_IP",        // Initially, valid modes are: HOST, BRIDGE, and CONTAINER_IP.
                                     // The default mode is HOST.
                                     // BRIDGE: only supported when using the Docker containerizer for now.
                                     // CONTAINER_IP: requires IP addresses to be included in Mesos resource offers.
      "protocolVersion": 4,          // Q: Is the protocol version necessary, or can we safely assume IPv4 for now?
      "sticky": false                // The `sticky` value should default to `false`.
                                     // Q: How would this be enforced?
    },
    "discovery": {
      "visibility": "EXTERNAL",      // The default visibility mode is CLUSTER.
      "name": "...",                 // Name, environment, location, version, and labels are optional.
      "environment": "...",          // These values may be used later by external systems like Mesos-DNS.
      "location": "...",
      "version": "...",
      "labels": {
        "one": "aaa",
        "two": "bbb"
      }
    },
    "netGroups": ["abc", "def"],     // These are network groups to which this app's tasks belong.
                                     // Network groups form the basis for whitelisting inbound connections.
                                     // See `.network.namedPorts[].allowedNetGroups` for an example.
    "namedPorts": [
      {
        "name": "http",
        "protocols": ["tcp", "udp"], // Q: Is it a problem that the Mesos DiscoveryInfo protobuf message only allows one protocol per port?
                                     //    Multiple protocols can be encoded in a single string if necessary.
        "allowedNetGroups": ["xyz"]  // Describes allowed inbound connections base on network groups
                                     //   - Inbound connections from within the same network group are always allowed.
                                     //   - By default, all inbound connections are allowed.
        "hostPort": 0,               // Host port has the traditinal meaning in Marathon:
                                     //   - If the value is set to '0', claim any available port from the Mesos resource offer.
                                     //   - If the value is a positive integer, then wait for a resource offer that includes that exact port.
        "containerPort": 80,         // Container port is only allowed with Docker bridge mode.
        "servicePort": 9001          // Service port has the traditional meaning in Marathon, to support discovery based on well-known ports.
      }
    ]
  }
}
```

Looking forward to comments!

**Comment by cmaloney** [comment]
As far as protocolVersion: There are several people / groups actively hoping to do ipv6 stuff, so protocolVersion seems useful, althuogh maybe I'd make it `'ipv6': False` rather than protocolVersion since there are a lot of "invalid" inputs people can give for that (It also doesn't tell me what I'm toggling very much).

How does netGroups interact with defining networks via roles in mesos? In particular how does this tie into / interact with #1460?

Is servicePort still required / necessary? If someone omits it can they not get assigned one automatically anymore?

**Comment by kozyraki** [comment]
We should take a look at [docker's libnetwork design] (https://github.com/docker/libnetwork/blob/master/docs/design.md) to make sure we are not missing something.

**Comment by BenWhitehead** [comment]
Nice proposal @ConnorDoyle. A couple questions:

1. What does `.network.addresses.sticky` mean?
2. Am I correct in assuming that `.network.discovery.visibility` is applied to the whole container? Seems like it might be desirable to configure my web app with an external port (443) but have a cluster only admin endpoint (7070).  Thought this could probably be addressed with being able to get multiple IP addresses and specify which to bind to.
3. Now that network information is being specified how does the task know which ip address to bind to?
4. I don't think we need `.network.namedPorts.protocols` to be an array. If the user needs multiple protocols for the same port number they should probably have a `namedPort` for that protocol. This also seems a little more logical to me since ports are protocol specific, so configuring `udp:80` would be different than `tcp:80`.  (can be tested with `nc -l localhost 5678` and `nc -u -l localhost 5678`)

I tend to agree with cody that we probably want to have a flag that simply `ipv6`, iptables also take this approach `--ipv6` vs `--ipv4`.

**Comment by ConnorDoyle** [comment]
#### IP version

@cmaloney, @BenWhitehead: How about an enumeration where the only valid values are `IPv4` (default) and `IPv6`?

#### Service ports

@cmaloney: They won't be required.  I'd like to _not_ assign one by default and make it opt-in.  What's your preference?

#### Address Stickiness

@BenWhitehead: This is a concept related to dynamic reservations in Mesos.  That is, in the process of launching a task, also name this resource (an address in this case) and offer it back to this framework first when the task terminates.  This is all dependent on how the resource might be represented in the future, and also whether global resources (those not associated with a particular host) will be integrated right away with the dynamic reservations system in Mesos.  Realistically it probably wouldn't make it into the first version of this API.

#### Discoverability

@BenWhitehead: Networking isolation, like other isolators in Mesos, will be done per-container.  The current DiscoveryInfo message is also defined for a task, but that won't help very much in Marathon where tasks and executor instances are by de-facto in 1:1 correspondence.  If in the future Mesos is updated to allow a visibility parameter on a named port, we can update Marathon to support it.

#### NamedPort Protocol Array

@BenWhitehead: Good call, we can send more named port elements to describe multiple protocols.

**Comment by ConnorDoyle** [comment]
This issue is blocked by #1505.

**Comment by cmaloney** [comment]
#### IP Version
The enumeration works for me, although I should ideally be able to specify both IPv4 and IPv6 at the same time.

#### Service Ports
I'd rather not assign one and have it be opt-in like you suggest. I suspect that one is always given will be a scaling limitation shortly with some of the other marathon work which has been going on

**Comment by BenWhitehead** [comment]
I think and enumeration could work as long as the error message was explicit in defining the possible values.  Would the parsing of the values be case insensitive or strict?

Service ports being optional sounds good to me.

Stickiness makes sense with that explanation, thanks Connor.

Discoverability mapping directly to container isolators makes sense, though we will need to be explicit in explaining that it is applied to the container (this will become more important as more frameworks start tying to leverage the IP resources).

**Comment by cmaloney** [comment]
I had a use case which came up recently where I'd need more than just the functionality porosed / discussed so far.

I'm trying to launch an exhibitor zookeeper cluster inside of a Mesos cluster. I need to have 3 instances of the exhibitor app. Each instance of the app has 4 ports. One for the exhibitor UI (http), and three for zookeeper (client, connect, and leader).

For the ports, the exhibitor port can be anything, and vary between the different instances of the app. For the other three ports, they need to be consistent across instances, but otherwise can be arbitrary.

Exhibitor needs to be able to generate a config file for zookeeper which matches on all the different hosts, and its current logic to do so requires that all the zookeeper ports be the same.

Ideally there is some way for me to ask marathon to assign an arbitrary port but make it the same across all instances of the app.

**Comment by kolloch** [comment]
There is a general related concept which is on my mind lately, though, which I would like everyone to consider. From what I understood from Kubernetes, they separate the endpoint configuration from launching the tasks. I very much like the idea.

It means that you can create an endpoint e.g. port 80 on a dynamically chosen IP which will stay fixed afterwards first. Afterwards you create the apps and you change your endpoint to include those apps (for example via labels). This allows you to restart/drop your apps or completely change how you start the apps for a certain service and at the same time keep the endpoint unchanged.

That of course also means that all your communication has to go to one level of indirection. That can be implemented with haproxy, iptables, advanced routing whatever but of course can lead to problems especially with applications which auto-discover their own IP address.

In most cases though, the app configuration could just use dynamic ports without any service ports or special configuration.

Even if we still need to cover the special cases, I would maybe separate service discovery options into an "endpoints" REST endpoint and only put the rest into the app definition.

What do you think?

**Comment by cmaloney** [comment]
I would urge strongly staying away from any sort of strong decision on what the Network looks like, that we need to setup magic IP address, ports, loadbalancers, etc. They are not going to be acceptable in a lot of environments, and the methods that, for instance, Kubernetes uses to do it work well in Google's infrastructure but not many others.

It would be much cleaner in my opinion to let people build up those networks themselves on top of the base apps that Marathon provides. For example, if I can get an IP per Task as has been discussed, then I can launch a marathon app with a set / fixed IP. That app is a load balancer, which can forward to the final machines which have arbitrary IPs / hostnames. The LoadBalancer app is a component of the overall app. All the infrastructure is nicely visible to administrators of the app.

Encoding knowledge of "the network needs to be setup like this" is going to break a lot of users, and make it a lot more complicated to get a basic marathon up and running. There is a lot of value in that you can currently take a marathon.jar, and easily run it wherever to add Marathon to a Mesos cluster.

**Comment by BenWhitehead** [comment]
I agree with @cmaloney I think that we should make sure to keep the goal of marathon and load balancing separate.  Kubernetes is a fair amount of magic that I don't think we can expect to work in all environments or scenarios.

For example I won't want to be connecting to a load balanced "service ip" for a service with a stateful protocol (i.e. cassandra/kafka) I will need to connect to each of the tasks directly.

**Comment by ConnorDoyle** [comment]
@BenWhitehead and @cmaloney I am inclined to agree.  From recent discussions it looks like the way forward for Mesos is that although there might be a first-class representation of addresses and subnets, they will likely _not_ be modeled as "global resources".  As you both point out, there are several related but unique aspects that could/should all be managed separately.

1. IP Address Management.
1. Interaction with per-host network isolation mechanisms.
1. Overlay policies.
1. Service discovery.

To this end, here are some general requirements for Mesos networking:

- Acquire IP address on-demand by the time a task launch happens on the Mesos slave.
- Framework can express intent to assign IP addresses when launching tasks.
- Communicate assigned IP address(es) to the scheduler.
- Expose assigned IP addresses via the Mesos state endpoint.
- Release IP address(es) when tasks complete.
- Network isolation strategy should be flexible enough to allow other systems (critically, Kubernetes) to hook into the IP allocation path.

@cmaloney is it a deal-breaker if you can't choose a specific address for a task?  This is probably not the correct forum for such a discussion.  @karya0 is working on an updated proposal.  Will link it here as soon as it's ready.

**Comment by ConnorDoyle** [comment]
Renamed this issue to be a larger umbrella for networking API, created sub-task to match the original intent.

**Comment by ConnorDoyle** [comment]
Related Mesos JIRA [here](https://issues.apache.org/jira/browse/MESOS-3013) (add a `NetworkInfo` message).

**Comment by ShimiTaNaka** [comment]
Any idea when this task will be completed?
Mesos is now 0.25 and the DiscoveryInfo is still not visible for the state.json endpoint. Very limiting for service discovery

**Comment by meichstedt** [comment]
Note: This issue has been migrated to https://jira.mesosphere.com/browse/MARATHON-4091. For more information see https://groups.google.com/forum/#!topic/marathon-framework/khtvf-ifnp8.
